core/drivers/amd/mailbox_driver.c

```diff
@@ -0,0 +1,212 @@
+					    rsp_addr,
+					    payload_size);
+	if (!ipi_info.req || !ipi_info.rsp)
+		panic();
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_call(IPI_MAILBOX_OPEN, IPI_BLOCK);
+	if (res != TEE_SUCCESS)
+		EMSG("Failed to open Mailbox for remote id %d", remote_id);
+
+	mutex_unlock(&ipi_info.lock);
+
+	return res;
+}
+
+TEE_Result mailbox_release(uint32_t remote_id, size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (ipi_info.remote != remote_id) {
+		EMSG("Mailbox not available for remote id %d", remote_id);
+		return res;
+	}
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_call(IPI_MAILBOX_RELEASE, IPI_BLOCK);
+	if (res != TEE_SUCCESS) {
+		EMSG("Mailbox release failed for remote id %d", remote_id);
+		goto err;
+	}
+
+	res = core_mmu_remove_mapping(MEM_AREA_RAM_SEC,
+				      ipi_info.req,
+				      payload_size);
+	res |= core_mmu_remove_mapping(MEM_AREA_RAM_SEC,
+				       ipi_info.rsp,
+				       payload_size);
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to remove memory mappings for remote id %d",
+		     remote_id);
+	}
+
+err:
+	mutex_unlock(&ipi_info.lock);
+	return res;
+}
+
+static TEE_Result mailbox_write(uint32_t remote_id,
+				void *payload,
+				size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (ipi_info.remote != remote_id) {
+		EMSG("Mailbox not available for remote id %d", remote_id);
+		goto err;
+	}
+
+	if (!IS_ALIGNED((uintptr_t)payload, CACHELINE_LEN)) {
+		EMSG("payload address not aligned");
+		goto err;
+	}
+
+	if (!IS_ALIGNED(payload_size, CACHELINE_LEN)) {
+		EMSG("payload size not aligned");
+		goto err;
+	}
+
+	memcpy(ipi_info.req, payload, payload_size);
+
+	cache_operation(TEE_CACHEFLUSH, ipi_info.req, payload_size);
+
+	res = TEE_SUCCESS;
+err:
+	return res;
+}
+
+static TEE_Result mailbox_read(uint32_t remote_id,
+			       void *payload,
+			       size_t payload_size,
+			       uint32_t *status)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cache_operation(TEE_CACHEINVALIDATE, ipi_info.rsp, payload_size);
+
+	*status = *(uint32_t *)ipi_info.rsp;
+
+	if (*status)
+		goto err;
+
+	if (ipi_info.rsp) {
+		memcpy(ipi_info.rsp, payload, payload_size);
+		res = TEE_SUCCESS;
+	}
+
+err:
+	return res;
+}
+
+TEE_Result mailbox_notify(uint32_t remote_id,
+			  void *payload,
+			  size_t payload_size,
+			  uint32_t blocking)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t remote_status = 0;
+
+	if (ipi_info.remote != remote_id) {
+		EMSG("Mailbox not available for remote id %d", remote_id);
+		return res;
+	}
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_write(ipi_info.remote, payload, payload_size);
+	if (res) {
+		EMSG("Can't write the request command");
+		goto out;
+	}
+
+	res = mailbox_call(IPI_MAILBOX_NOTIFY, blocking);
+	if (res) {
+		EMSG("IPI error");
+		goto out;
+	}
+
+	res = mailbox_read(ipi_info.remote, payload, payload_size, &remote_status);
+	if (res)
+		EMSG("Can't read the remote response");
+
+	if (remote_status)
```

Comment:
Nitpicking: `res` is already set to `TEE_ERROR_GENERIC` when `remote_status` is not 0.